From 806a3203ba363432a7f5852795aa136e32ca3006 Mon Sep 17 00:00:00 2001
From: Sunil Goutham <sgoutham@marvell.com>
Date: Mon, 13 Aug 2018 16:03:18 +0530
Subject: [PATCH 0120/1051] net: octeontx2: Set skb->hash from HW

If rxhash offload is enabled then set skb's hash
to flow_tag from CQE Rx descriptor.

Signed-off-by: Sunil Goutham <sgoutham@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../ethernet/marvell/octeontx2/otx2_common.c  |  1 +
 .../net/ethernet/marvell/octeontx2/otx2_pf.c  |  4 +--
 .../ethernet/marvell/octeontx2/otx2_txrx.c    | 25 +++++++++++++++++++
 3 files changed, 28 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/otx2_common.c b/drivers/net/ethernet/marvell/octeontx2/otx2_common.c
index 32c6671a2390..718ee3d8ec20 100644
--- a/drivers/net/ethernet/marvell/octeontx2/otx2_common.c
+++ b/drivers/net/ethernet/marvell/octeontx2/otx2_common.c
@@ -416,6 +416,7 @@ static int otx2_rq_init(struct otx2_nic *pfvf, u16 qidx)
 	aq->rq.lpb_aura = qidx; /* Use large packet buffer aura */
 	aq->rq.lpb_sizem1 = (DMA_BUFFER_LEN / 8) - 1;
 	aq->rq.xqe_imm_size = 0; /* Copying of packet to CQE not needed */
+	aq->rq.flow_tagw = 32; /* Copy full 32bit flow_tag to CQE header */
 
 	/* Fill AQ info */
 	aq->qidx = qidx;
diff --git a/drivers/net/ethernet/marvell/octeontx2/otx2_pf.c b/drivers/net/ethernet/marvell/octeontx2/otx2_pf.c
index 5cd3fcfe93e8..13731f5f3686 100644
--- a/drivers/net/ethernet/marvell/octeontx2/otx2_pf.c
+++ b/drivers/net/ethernet/marvell/octeontx2/otx2_pf.c
@@ -737,8 +737,8 @@ static int otx2_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 	 */
 	pf->iommu_domain = iommu_get_domain_for_dev(dev);
 
-	netdev->hw_features = (NETIF_F_RXCSUM |
-			       NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM);
+	netdev->hw_features = (NETIF_F_RXCSUM | NETIF_F_IP_CSUM |
+			       NETIF_F_IPV6_CSUM | NETIF_F_RXHASH);
 	netdev->features |= netdev->hw_features;
 
 	netdev->netdev_ops = &otx2_netdev_ops;
diff --git a/drivers/net/ethernet/marvell/octeontx2/otx2_txrx.c b/drivers/net/ethernet/marvell/octeontx2/otx2_txrx.c
index 5d35176d0689..e70d1f63bf83 100644
--- a/drivers/net/ethernet/marvell/octeontx2/otx2_txrx.c
+++ b/drivers/net/ethernet/marvell/octeontx2/otx2_txrx.c
@@ -112,6 +112,29 @@ static void otx2_snd_pkt_handler(struct otx2_nic *pfvf,
 	}
 }
 
+static inline void otx2_set_rxhash(struct otx2_nic *pfvf,
+				   struct nix_cqe_hdr_s *cqe_hdr,
+				   struct sk_buff *skb)
+{
+	enum pkt_hash_types hash_type = PKT_HASH_TYPE_NONE;
+	struct otx2_rss_info *rss;
+	u32 hash = 0;
+
+	if (!(pfvf->netdev->features & NETIF_F_RXHASH))
+		return;
+
+	rss = &pfvf->hw.rss_info;
+	if (rss->flowkey_cfg) {
+		if (rss->flowkey_cfg &
+		    ~(FLOW_KEY_TYPE_IPV4 | FLOW_KEY_TYPE_IPV6))
+			hash_type = PKT_HASH_TYPE_L4;
+		else
+			hash_type = PKT_HASH_TYPE_L3;
+		hash = cqe_hdr->flow_tag;
+	}
+	skb_set_hash(skb, hash, hash_type);
+}
+
 static void otx2_skb_add_frag(struct otx2_nic *pfvf,
 			      struct sk_buff *skb, u64 iova, int len)
 {
@@ -222,6 +245,8 @@ static void otx2_rcv_pkt_handler(struct otx2_nic *pfvf,
 	if (!skb)
 		return;
 
+	otx2_set_rxhash(pfvf, cqe_hdr, skb);
+
 	skb_record_rx_queue(skb, cq->cq_idx);
 	skb->protocol = eth_type_trans(skb, pfvf->netdev);
 	if (pfvf->netdev->features & NETIF_F_RXCSUM)
-- 
2.17.1

